arch/x86/core/pcie.c

```diff
@@ -217,8 +217,8 @@ uint16_t pcie_msi_mdr(unsigned int irq,
 		return 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
 	}
 
-	/* VT-D requires it to be 0, so let's return 0 by default */
-	return 0;
+	/* VT-D requires sub handle and msi it is vector */
+	return vector->arch.vector;
```

Comment:
No MSI data register is meant to be 0. Check the VT-D specs section 5.1.5.2

Reply:
Please read full spec. There are following option as per the spec for configure MSI add/data registers and it is not true that MSI data register meant to be 0.
SHV = 0; handle = interrupt indx, here you can use MSI Address register for interrupt index
SHV = 1; handle = interrupt indx; subhandle = 0, here also you can use MSI Address register for interrupt index
SHV = 1; handle = 0; subhandle = interrupt indx, here you can use data register for interrupt index
SHV = 1; handle = interrupt index - subhandle, here can use both address and data register for indexing

Reply:
Looks like you are confusing how hardware must create the MSI message prior to how the software has configured it.

Reply:
sorry this a coding mistake which I suppose to return vector->arch.irte (interrupt index) instead of vector. But please note that we can use MSI data register for interrupt index as sub handle and it is not true that MSI data register meant to be 0. But only issue in that case is interrupt index should aligned to Multiple Message Capable bits (number of vector requested by the device).

Reply:
no, it's not vector or irte: it's 0. Or then specs would be really wrong (to be fair, there are inaccuracies and most of it is sometime really confusing)

The handle/sub-handle logic you described (and which is also described in section 5.1.5, though I agree: it's badly written and can lead to confusion) is meant to be implemented by the MSI device, not the host configuration software, which is zephyr here. 
So let's say device D is MSI and needs 4 vectors (announced in MMC) (vector 1 is 0, vector 2 is 1, vector 3 is 2 and vector 4 is 3)
zephyr: programs device D MSI MME so it enables the 4 vectors, in its address register it tells the first interrupt index of the related irte of first 4 vectors (let's call it base_irte), SHV to 1, interrupt format to 1 of course and data register is always 0.
Device D: fires its vector 2 for instance so it  generatest an MSI message with an address/data register pair so:
- adresss owns base_irte + 1, data owns 0: vt-d will take the handle as interrupt_index
- adress owns 0, data owns base_irte + 1: vt-d will take the sub-handle as interrupt_index
- address owns (base_irte + 1) - x, data owns x: vt-d will compute handle + sub_handle as interrupt index. (note that this formula explains also the 2 cases above if x = 0 and x = base_irte + 1)

Look, if you don't trust me (you definitely don't), perhaps you'll trust ACRN source code? Search for ptirq_build_physical_msi().